internal/statesync/reactor.go

```diff
@@ -237,8 +214,89 @@ func NewReactor(
 // The caller must be sure to execute OnStop to ensure the outbound p2p Channels are
 // closed. No error is returned.
 func (r *Reactor) OnStart(ctx context.Context) error {
-	go r.processChannels(ctx, r.snapshotCh, r.chunkCh, r.blockCh, r.paramsCh)
-	go r.processPeerUpdates(ctx)
+	// construct channels
+	chDesc := getChannelDescriptors()
+	snapshotCh, err := r.chCreator(ctx, chDesc[SnapshotChannel])
+	if err != nil {
+		return err
+	}
+	chunkCh, err := r.chCreator(ctx, chDesc[ChunkChannel])
+	if err != nil {
+		return err
+	}
+	blockCh, err := r.chCreator(ctx, chDesc[LightBlockChannel])
+	if err != nil {
+		return err
+	}
+	paramsCh, err := r.chCreator(ctx, chDesc[ParamsChannel])
+	if err != nil {
+		return err
+	}
+
+	// define constructor and helper functions, that hold
+	// references to these channels for use later. This is not
+	// ideal.
```

Comment:
fair, we still have to define it in the scope of this function so it can access the channels. we could get rid of the `newSyncer` constructor.

Reply:
> fair, we still have to define it in the scope of this function so it can access the channels. we could get rid of the `newSyncer` constructor.

Yeah, I only meant that specific one—I realize in retrospect I put my comment on the comment instead of the code line, implying more generality than I intended.